---
layout: post
title: "Section 5: Asynchronous JavaScript: Promises and Async/Await"
date: 2024-12-30
draft: false
categories:
  - Computer
tags:
  - Node.js
---

# Asynchronous JavaScript: Promises and Async/Await

`JavaScript` is asynchronous, and it relies on call-back functions to build its architecture. However, modern `JavaScript` has been more and more written using new tools, like "promises" and "async/await", to avoid "call-back hell".

## Super Agent

It's a module that is good at retrieving data from APIs. Here is an example of how to use it. Of course, here the `superagent` is nested in a `fs.readFile` call-back function, and `superaent` itself is nesting a `fs.writeFile` function.

```js
const superagent = require("superagent");

fs.readFile(`${__dirname}/dog.txt`, (error, data) => {
	superagent
		.get(`https://dog.ceo/api/breed/${data}/images/random`)
		.end((error, result) => {
			if (error) console.log(error.message);
			fs.writeFile("dog-img.txt", result.body.message, (error) => {
				if (error) {
					console.log(error.message);
				} else {
					console.log("Dog image file path saved!");
				}
			});
		});
});
```

## A Promise Example

`superagent` comes with a nice promise method, avoiding the call back hell situation.

Please note the `.then` method deals with successful situation and `.catch` deals with failed situation individually.

```js
superagent
	.get(`https://dog.ceo/api/breed/${data}/images/random`)
	.then((result) => {
		fs.writeFile("dog-img.txt", result.body.message, (error) => {
			if (error) {
				console.log(error.message);
			} else {
				console.log("Dog image file path saved!");
			}
		});
	})
	.catch((error) => {
		console.log(error.message);
	});
```

## Build Your Own Promises

We can build promises for `readFile` and `writeFile`:

```js
function readFilePromise(filePath) {
	return new Promise((resolve, reject) => {
		fs.readFile(filePath, (error, data) => {
			if (error) reject("File not found!");
			resolve(data);
		});
	});
}
```

```js
function writeFilePromise(filePath, data) {
	return new Promise((resolve, reject) => {
		fs.writeFile(filePath, data, (error) => {
			if (error) reject("Could not save file!");
			resolve("File saved successfully.");
		});
	});
}
```

`Promise` is a new class introduced into `JavaScript` since `ES6`. Please note `resolve` and `reject` are just common names referred to success and failure situations. You can use any other words instead.

With all these promises, we can re-write the code as follows.

```js
readFilePromise(`${__dirname}/dog.txt`)
	.then((data) => {
		console.log(`Breed: ${data}`);
		return superagent.get(`https://dog.ceo/api/breed/${data}/images/random`);
	})
	.then((result) => {
		console.log(result.body.message);
		return writeFilePromise("dog-img.txt", result.body.message);
	})
	.then((result) => {
		console.log(result);
	})
	.catch((error) => console.log(error.message));
```

The key point is to use `return` every time to trigger the next function.

<p class="highlight yellow" markdown="1">`.catch` here is not attached to the last function, but to every function before it.</p>

## Async/Await

`Async/Await` was introduced into `ES8`. It is a further step beyond `.then`. There is the re-coding using `Async/Await` for the above example.

```js
async function getDocPic() {
	const data = await readFilePromise(`${__dirname}/dog.txt`);
	console.log(`Breed: ${data}`);

	const result = await superagent.get(
		`https://dog.ceo/api/breed/${data}/images/random`
	);
	console.log(result.body.message);

	const result2 = await writeFilePromise("dog-img.txt", result.body.message);
	console.log(result2);
}

getDocPic();
```

We haven't dealt with error situation in the above code. But it's easy, as we just need to wrap the main content into a `try-catch` bracket.

```js
try {
	// main body of the codes
} catch (error) {
	console.log(error.message);
}
```

## Returning Values from Async Functions

We **cannot** just store a returned value from async function by `const x = asyncFunc()`, because that will only give us a `promise (pending)` result.

We should use `.then` method:

```js
asyncFunc().then((result) => {
	const x = result;
});
```

<p class="highlight yellow" markdown="1">Asynchronous function won't return an error automatically. You need to use `throw` method.</p>

There is a example of `throw` method.

```js
try {
	// main body of the codes
} catch (error) {
	console.log(error);
	throw error;
}
```

And here is how the error message is used.

```js
getDogPic()
	.then((result) => {
		console.log(result);
	})
	.catch((error) => console.log(error));
```

We are having the `.then` and `.catch` methods again. But you can still use async function paradigm to implement the function itself.

```js
async () => {
	try {
		await getDogPic();
	} catch (error) {
		console.log(error);
	}
};
```

However, if you are uncomfortable with this way of implementing async function, you are absolute free to go back to the `.then` and `.catch` methods.

## Multiple Promises Simultaneously

If we have multiple promises to implement at the same time, don't use `await async function` one by one, because that will block the progress. we can use `Promise.all([p1,p2,p3])` instead.

For example, if we want to implement `superagent.get()` three times, we have the following code.

```js
const allRes = await Promise.all([
	superagent.get(`https://dog.ceo/api/breed/${data}/images/random`),
	superagent.get(`https://dog.ceo/api/breed/${data}/images/random`),
	superagent.get(`https://dog.ceo/api/breed/${data}/images/random`),
]);
//const imgPaths = allRes.map((element) => element.body.message);
```

A more common way will be storing each promise in a variable and implement it together with other promises.

```js
const res1 = superagent.get(`https://dog.ceo/api/breed/${data}/images/random`);
const res2 = superagent.get(`https://dog.ceo/api/breed/${data}/images/random`);
const res3 = superagent.get(`https://dog.ceo/api/breed/${data}/images/random`);

const allRes = await Promise.all([res1, res2, res3]);
```
